src/poetry/utils/authenticator.py

```diff
@@ -104,10 +109,26 @@ def http_credential_keys(self) -> list[str]:
     def get_http_credentials(
         self, password_manager: PasswordManager, username: str | None = None
     ) -> HTTPAuthCredential:
-        # try with the repository name via the password manager
-        credential = HTTPAuthCredential(
-            **(password_manager.get_http_auth(self.name) or {})
-        )
+        # try with the URL
+        parsed_url = urllib.parse.urlsplit(self.url)
+        if "@" in parsed_url.netloc:
+            # Split from the right because that's how urllib.parse.urlsplit()
+            # behaves if more than one @ is present (which can be checked using
+            # the password attribute of urlsplit()'s return value).
+            auth, netloc = parsed_url.netloc.rsplit("@", 1)
+            # Split from the left because that's how urllib.parse.urlsplit()
+            # behaves if more than one : is present (which again can be checked
+            # using the password attribute of the return value)
+            user, password = auth.split(":", 1) if ":" in auth else (auth, "")
+            credential = HTTPAuthCredential(user, password)
+        else:
+            credential = HTTPAuthCredential(None, None)
+
+        if credential.username is None and credential.password is None:
+            # try with the repository name via the password manager
+            credential = HTTPAuthCredential(
+                **(password_manager.get_http_auth(self.name) or {})
+            )
```

Comment:
```suggestion
        else:
            # try with the repository name via the password manager
            credential = HTTPAuthCredential(
                **(password_manager.get_http_auth(self.name) or {})
            )
```
No need to create a new `HTTPAuthCredential(None, None)`